os/binfmt/libelf/libelf_cache.c

```diff
@@ -0,0 +1,502 @@
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: elf_cache_blocks_to_read
+ *
+ * Description:
+ *   Assign values to first_block, last_block and number of blocks to be read
+ *   from elf file in this particular call to elf_cache_read
+ *
+ * Returned Value:
+ *   None
+ ****************************************************************************/
+static void elf_cache_blocks_to_read(int *first_block, int *last_block, int *no_blocks, int offset, int readsize)
+{
+	int blocksize;
+
+	blocksize = cache_blocks_size;
+
+	*first_block = offset / blocksize;
+	*last_block = (offset + readsize) / blocksize;
+	*no_blocks = *last_block - *first_block + 1;
+}
+
+/****************************************************************************
+ * Name: elf_cache_lseek_block
+ *
+ * Description:
+ *   Seek to start of "block_number" block in elf blocks section
+ *
+ * Returned Value:
+ *   Starting offset (positive) of 'block_number' block in binary on Success
+ *   Negative value on Failure
+ ****************************************************************************/
+static off_t elf_cache_lseek_block(int filfd, uint16_t binary_header_size, int block_number)
+{
+	off_t rpos;
+	off_t actual_offset;
+
+	actual_offset = binary_header_size + block_number * cache_blocks_size;;
+
+	/* Seek to location of this block in actual ELF file */
+	rpos = lseek(filfd, actual_offset, SEEK_SET);
+
+	binfo("filfd: %d actual_offset: %d rpos: %d\n", filfd, actual_offset, rpos);
+
+	if (rpos != actual_offset) {
+		int errval = get_errno();
+		berr("Failed to seek to position %lu: %d\n", (unsigned long)actual_offset, errval);
+		return -errval;
+	}
+
+	return rpos;
+}
+
+/****************************************************************************
+ * Name: elf_cache_read_block
+ *
+ * Description:
+ *   Read 'block_number' block from elf blocks section into
+ *   provided respective block numbers 'out_buffer' variable.
+ *
+ * Returned Value:
+ *   Number of bytes read into out_buffer on Success
+ *   Negative value on Failure
+ ****************************************************************************/
+static off_t elf_cache_read_block(int filfd, uint16_t binary_header_size, FAR uint8_t *buf, int block_number)
+{
+	off_t rpos;
+	size_t readsize;
+	ssize_t nbytes;
+
+	binfo("filfd: %d block_number: %d binary_header_size: %d\n", filfd, block_number, binary_header_size);
+
+	/* Seek to location of 'block_number' block in elf file */
+	rpos = elf_cache_lseek_block(filfd, binary_header_size, block_number);
+
+	if (rpos < 0) {
+		berr("Failed to seek to offset of block number %d\n", block_number);
+		return rpos;
+	}
+
+	/* Last unaligned blocks to be read with its actual size and not with blocksize;*/
+	if (block_number == number_of_blocks - 1) {
```

Comment:
can you make both as unsigned if there are no negative values expected, it is mismatched here